distributed/tests/test_scheduler.py

```diff
@@ -3183,3 +3183,25 @@ async def test_worker_heartbeat_after_cancel(c, s, *workers):
 
     while any(w.tasks for w in workers):
         await asyncio.gather(*[w.heartbeat() for w in workers])
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_worker_reconnect_task_memory_with_resources(c, s, a):
+    b = await Worker(s.address, resources={"A": 1})
+    b.periodic_callbacks["heartbeat"].stop()
+
+    futs = c.map(inc, range(10), resources={"A": 1})
+    res = c.submit(sum, futs)
+
+    while not b.executing_count and not b.data:
+        await asyncio.sleep(0.001)
+
+    await s.remove_worker(address=b.address, close=False)
+    while not res.done():
+        await b.heartbeat()
+
+    await res
+    assert ("no-worker", "memory") in {
+        (start, finish) for (_, start, finish, _, _) in s.transition_log
+    }
+    await b.close()
```

Comment:
Thanks @fjetter!

One small comment, can we use a context manager for `Worker` instead of explicitly calling `close`?